python/ray/air/_internal/tensorflow_utils.py

```diff
@@ -10,21 +10,24 @@
     from ray.data._internal.pandas_block import PandasBlockSchema
 
 
-def convert_ndarray_to_tf_tensor(
-    ndarray: np.ndarray,
+def convert_ndarray_batch_to_tf_tensor(
+    batch: np.ndarray,
     dtype: Optional[tf.dtypes.DType] = None,
 ) -> tf.Tensor:
     """Convert a NumPy ndarray to a TensorFlow Tensor.
 
     Args:
-        ndarray: A NumPy ndarray that we wish to convert to a TensorFlow Tensor.
+        batch: A NumPy ndarray that we wish to convert to a TensorFlow Tensor.
         dtype: A TensorFlow dtype for the created tensor; if None, the dtype will be
             inferred from the NumPy ndarray data.
 
     Returns: A TensorFlow Tensor.
     """
-    ndarray = _unwrap_ndarray_object_type_if_needed(ndarray)
-    return tf.convert_to_tensor(ndarray, dtype=dtype)
+    batch = _unwrap_ndarray_object_type_if_needed(batch)
+    if batch.dtype == object:
```

Comment:
I don't think that this is a reliable identifier for a ragged tensor, certain e.g. string-typed ndarrays could have an object dtype I believe. I think that the only way to reliably determine if it's a ragged tensor is to check that it has an object dtype **and** to check whether at least one element-ndarray has a different shape that the others. We do that explicit check here, for example: https://github.com/ray-project/ray/blob/d08cddc351eedc4a3af1fadb6e509189beabb756/python/ray/air/util/tensor_extensions/utils.py#L4-L22

Reply:
A cleaner solution that wouldn't require O(#rows) checks on this conversion would be to try to `tf.convert_to_tensor()` and fallback to `tf.ragged.constant()` on a particular exception type/messsage that's thrown, keeping the common case (homogeneous-shaped tensors) fast. We do that for [the equivalent Torch tensor conversion](https://github.com/ray-project/ray/blob/d08cddc351eedc4a3af1fadb6e509189beabb756/python/ray/air/_internal/torch_utils.py#L58-L81), as well as when [constructing the tensor extension array](https://github.com/ray-project/ray/blob/d08cddc351eedc4a3af1fadb6e509189beabb756/python/ray/air/util/tensor_extensions/pandas.py#L1424-L1448).

What error does `tf.convert_to_tensor()` throw when given a variable-shaped ndarray?

Reply:
> What error does tf.convert_to_tensor() throw when given a variable-shaped ndarray?

You get:

```
ValueError: Failed to convert a NumPy array to a Tensor (Unsupported object type numpy.ndarray).
```

I've updated the implementation to infer raggedness from the type spec instead of from the ndarray dtype. See my other comment: https://github.com/ray-project/ray/pull/29476#discussion_r1016980130